File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.java
// Copyright 2015-present 650 Industries. All rights reserved.

package versioned.host.exp.exponent.modules.api.notifications;

import com.cronutils.model.Cron;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.google.firebase.iid.FirebaseInstanceId;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

import javax.inject.Inject;

import expo.modules.updates.manifest.raw.RawManifest;
import host.exp.exponent.Constants;
import host.exp.exponent.ExponentManifest;
import host.exp.exponent.analytics.EXL;
import host.exp.exponent.di.NativeModuleDepsProvider;
import host.exp.exponent.network.ExponentNetwork;
import host.exp.exponent.notifications.ExponentNotificationManager;
import host.exp.exponent.notifications.NotificationActionCenter;
import host.exp.exponent.notifications.NotificationConstants;
import host.exp.exponent.notifications.NotificationHelper;
import host.exp.exponent.notifications.exceptions.UnableToScheduleException;
import host.exp.exponent.notifications.managers.SchedulersManagerProxy;
import host.exp.exponent.notifications.schedulers.CalendarSchedulerModel;
import host.exp.exponent.notifications.schedulers.IntervalSchedulerModel;
import host.exp.exponent.notifications.schedulers.SchedulerImpl;
import host.exp.exponent.storage.ExponentSharedPreferences;

import static host.exp.exponent.notifications.helpers.ExpoCronParser.createCronInstance;

public class NotificationsModule extends ReactContextBaseJavaModule {

  private static final String TAG = NotificationsModule.class.getSimpleName();

  @Inject
  ExponentSharedPreferences mExponentSharedPreferences;

  @Inject
  ExponentManifest mExponentManifest;

  @Inject
  ExponentNetwork mExponentNetwork;

  private final RawManifest mManifest;

  public NotificationsModule(ReactApplicationContext reactContext,
                             RawManifest manifest, Map<String, Object> experienceProperties) {
    super(reactContext);
    NativeModuleDepsProvider.getInstance().inject(NotificationsModule.class, this);
    mManifest = manifest;
  }

  @Override
  public String getName() {
    return "ExponentNotifications";
  }

  @ReactMethod
  public void createCategoryAsync(final String categoryIdParam, final ReadableArray actions, final Promise promise) {
    String categoryId = getScopedIdIfNotDetached(categoryIdParam);
    List<Map<String, Object>> newActions = new ArrayList<>();

    for (Object actionObject : actions.toArrayList()) {
      if (actionObject instanceof Map) {
        Map<String, Object> action = (Map<String, Object>) actionObject;
        newActions.add(action);
      }
    }

    NotificationActionCenter.putCategory(categoryId, newActions);
    promise.resolve(null);
  }

  @ReactMethod
  public void deleteCategoryAsync(final String categoryIdParam, final Promise promise) {
    String categoryId = getScopedIdIfNotDetached(categoryIdParam);
    NotificationActionCenter.removeCategory(categoryId);
    promise.resolve(null);
  }

  private String getScopedIdIfNotDetached(String categoryId) {
    if (!Constants.isStandaloneApp()) {
      try {
        String experienceId = mManifest.getStableLegacyID();
        return experienceId + ":" + categoryId;
      } catch (JSONException e) {
        e.printStackTrace();
      }
    }
    return categoryId;
  }

  @ReactMethod
  public void getDevicePushTokenAsync(final ReadableMap config, final Promise promise) {
    if (!Constants.isStandaloneApp()) {
      promise.reject("getDevicePushTokenAsync is only accessible within standalone applications");
    }
    try {
      if (Constants.FCM_ENABLED) {
        String token = FirebaseInstanceId.getInstance().getToken();
        if (token == null) {
          promise.reject("FCM token has not been set");
        } else {
          WritableMap params = Arguments.createMap();
          params.putString("type", "fcm");
          params.putString("data", token);
          promise.resolve(params);
        }
      } else {
        promise.reject("ERR_NOTIFICATIONS_FCM_NOT_ENABLED", "FCM must be enabled in order to get the device push token");
      }
    } catch (Exception e) {
      EXL.e(TAG, e.getMessage());
      promise.reject(e.getMessage());
    }
  }

  @ReactMethod
  public void getExponentPushTokenAsync(final Promise promise) {
    String uuid = mExponentSharedPreferences.getUUID();
    if (uuid == null) {
      // This should have been set by ExponentNotificationIntentService via
      // ExpoFcmMessagingService#onNewToken -> FcmRegistrationIntentService.registerForeground -> ExponentNotificationIntentService#onHandleIntent
      // (#onNewToken is supposed to be called when a token is generated after app install, see
      // https://developers.google.com/android/reference/com/google/firebase/messaging/FirebaseMessagingService#onNewToken(java.lang.String)).
      // If it hasn't been set, the app probably couldn't register at FCM (invalid configuration?).
      promise.reject("E_GET_PUSH_TOKEN_FAILED", "Couldn't get push token on device. Check that your FCM configuration is valid.");
      return;
    }

    try {
      String experienceId = mManifest.getStableLegacyID();
      NotificationHelper.getPushNotificationToken(uuid, experienceId, mExponentNetwork, mExponentSharedPreferences, new NotificationHelper.TokenListener() {
        @Override
        public void onSuccess(String token) {
          promise.resolve(token);
        }

        @Override
        public void onFailure(Exception e) {
          promise.reject("E_GET_PUSH_TOKEN_FAILED", "Couldn't get push token for device. Check that your FCM configuration is valid.", e);
        }
      });
    } catch (JSONException e) {
      promise.reject("E_GET_PUSH_TOKEN_FAILED", "Couldn't get push token for device. Check that your FCM configuration is valid.", e);
    }
  }

  @ReactMethod
  public void createChannel(String channelId, final ReadableMap data, final Promise promise) {
    String experienceId;
    String channelName;

    try {
      experienceId = mManifest.getStableLegacyID();
    } catch (Exception e) {
      promise.reject("E_FAILED_CREATING_CHANNEL", "Requires Experience ID");
      return;
    }

    if (data.hasKey(NotificationConstants.NOTIFICATION_CHANNEL_NAME)) {
      channelName = data.getString(NotificationConstants.NOTIFICATION_CHANNEL_NAME);
    } else {
      promise.reject("E_FAILED_CREATING_CHANNEL", "Requires channel name");
      return;
    }

    try {
      NotificationHelper.createChannel(
        getReactApplicationContext(),
        experienceId,
        channelId,
        channelName,
        data.toHashMap());
      promise.resolve(null);
    } catch (Exception e) {
      promise.reject("E_FAILED_CREATING_CHANNEL", "Could not create channel", e);
    }
  }

  @ReactMethod
  public void deleteChannel(String channelId, final Promise promise) {
    String experienceId;

    try {
      experienceId = mManifest.getStableLegacyID();
    } catch (Exception e) {
      promise.reject("E_FAILED_DELETING_CHANNEL", "Requires Experience ID");
      return;
    }

    try {
      NotificationHelper.deleteChannel(
        getReactApplicationContext(),
        experienceId,
        channelId);
      promise.resolve(null);
    } catch (Exception e) {
      promise.reject("E_FAILED_DELETING_CHANNEL", "Could not delete channel", e);
    }
  }

  @ReactMethod
  public void presentLocalNotification(final ReadableMap data, final Promise promise) {
    presentLocalNotificationWithChannel(data, null, promise);
  }

  @ReactMethod
  public void presentLocalNotificationWithChannel(final ReadableMap data, final ReadableMap legacyChannelData, final Promise promise) {
    HashMap<String, java.io.Serializable> details = new HashMap<>();
    String experienceId;

    HashMap<String, Object> hashMap = data.toHashMap();
    if (data.hasKey("categoryId")) {
      hashMap.put("categoryId", getScopedIdIfNotDetached(data.getString("categoryId")));
    }

    details.put("data", hashMap);

    try {
      experienceId = mManifest.getStableLegacyID();
      details.put("experienceId", experienceId);
    } catch (Exception e) {
      promise.reject("E_FAILED_PRESENTING_NOTIFICATION", "Requires Experience ID");
      return;
    }

    if (legacyChannelData != null) {
      String channelId = data.getString("channelId");
      if (channelId == null) {
        promise.reject("E_FAILED_PRESENTING_NOTIFICATION", "legacyChannelData was nonnull with no channelId");
        return;
      }
      NotificationHelper.maybeCreateLegacyStoredChannel(
        getReactApplicationContext(),
        experienceId,
        channelId,
        legacyChannelData.toHashMap());
    }

    int notificationId = new Random().nextInt();

    NotificationHelper.showNotification(
      getReactApplicationContext(),
      notificationId,
      details,
      mExponentManifest,
      new NotificationHelper.Listener() {
        public void onSuccess(int id) {
          promise.resolve(id);
        }

        public void onFailure(Exception e) {
          promise.reject(e);
        }
      });
  }

  @ReactMethod
  public void scheduleLocalNotification(final ReadableMap data, final ReadableMap options, final Promise promise) {
    scheduleLocalNotificationWithChannel(data, options, null, promise);
  }

  @ReactMethod
  public void scheduleLocalNotificationWithChannel(final ReadableMap data, final ReadableMap options, final ReadableMap legacyChannelData, final Promise promise) {
    if (legacyChannelData != null) {
      String experienceId;
      try {
        experienceId = mManifest.getStableLegacyID();
      } catch (JSONException e) {
        promise.reject("E_FAILED_PRESENTING_NOTIFICATION", "legacyChannelData was nonnull with no channelId or no experienceId");
        return;
      }

      String channelId = data.getString("channelId");
      if (channelId == null) {
        promise.reject("E_FAILED_PRESENTING_NOTIFICATION", "legacyChannelData was nonnull with no channelId or no experienceId");
        return;
      }
      NotificationHelper.maybeCreateLegacyStoredChannel(
        getReactApplicationContext(),
        experienceId,
        channelId,
        legacyChannelData.toHashMap());
    }

    int notificationId = new Random().nextInt();

    HashMap<String, Object> hashMap = data.toHashMap();
    if (data.hasKey("categoryId")) {
      hashMap.put("categoryId", getScopedIdIfNotDetached(data.getString("categoryId")));
    }

    NotificationHelper.scheduleLocalNotification(
      getReactApplicationContext(),
      notificationId,
      hashMap,
      options.toHashMap(),
      mManifest,
      new NotificationHelper.Listener() {
        public void onSuccess(int id) {
          promise.resolve(id);
        }

        public void onFailure(Exception e) {
          promise.reject(e);
        }
      });
  }

  @ReactMethod
  public void dismissNotification(final int notificationId, final Promise promise) {
    try {
      ExponentNotificationManager manager = new ExponentNotificationManager(getReactApplicationContext());
      manager.cancel(
        mManifest.getStableLegacyID(),
        notificationId
      );
      promise.resolve(true);
    } catch (JSONException e) {
      promise.reject(e);
    }
  }

  @ReactMethod
  public void dismissAllNotifications(final Promise promise) {
    try {
      ExponentNotificationManager manager = new ExponentNotificationManager(getReactApplicationContext());
      manager.cancelAll(mManifest.getStableLegacyID());
      promise.resolve(true);
    } catch (JSONException e) {
      promise.reject(e);
    }
  }

  @ReactMethod
  public void cancelScheduledNotificationAsync(final int notificationId, final Promise promise) {
    try {
      ExponentNotificationManager manager = new ExponentNotificationManager(getReactApplicationContext());
      manager.cancelScheduled(mManifest.getStableLegacyID(), notificationId);
      promise.resolve(null);
    } catch (Exception e) {
      promise.reject(e);
    }
  }

  @ReactMethod
  public void cancelScheduledNotificationWithStringIdAsync(final String id, final Promise promise) {
    try {
      SchedulersManagerProxy.getInstance(getReactApplicationContext()
        .getApplicationContext())
        .removeScheduler(id);
      promise.resolve(null);
    } catch (Exception e) {
      promise.reject(e);
    }
  }

  @ReactMethod
  public void cancelAllScheduledNotificationsAsync(final Promise promise) {
    try {
      ExponentNotificationManager manager = new ExponentNotificationManager(getReactApplicationContext());
      manager.cancelAllScheduled(mManifest.getStableLegacyID());

      String experienceId = mManifest.getStableLegacyID();

      SchedulersManagerProxy
        .getInstance(getReactApplicationContext().getApplicationContext())
        .removeAll(experienceId);

      promise.resolve(null);
    } catch (Exception e) {
      promise.reject(e);
    }
  }

  @ReactMethod
  public void scheduleNotificationWithTimer(final ReadableMap data, final ReadableMap optionsMap, final Promise promise) {
    HashMap<String, Object> options = optionsMap.toHashMap();
    int notificationId = Math.abs(new Random().nextInt());
    HashMap<String, Object> hashMap = data.toHashMap();
    if (data.hasKey("categoryId")) {
      hashMap.put("categoryId", getScopedIdIfNotDetached(data.getString("categoryId")));
    }
    HashMap<String, Object> details = new HashMap<>();
    details.put("data", hashMap);
    String experienceId;

    try {
      experienceId = mManifest.getStableLegacyID();
      details.put("experienceId", experienceId);
    } catch (Exception e) {
      promise.reject(new Exception("Requires Experience Id"));
      return;
    }

    IntervalSchedulerModel intervalSchedulerModel = new IntervalSchedulerModel();
    intervalSchedulerModel.setExperienceId(experienceId);
    intervalSchedulerModel.setNotificationId(notificationId);
    intervalSchedulerModel.setDetails(details);
    intervalSchedulerModel.setRepeat(options.containsKey("repeat") && (Boolean) options.get("repeat"));
    intervalSchedulerModel.setScheduledTime(System.currentTimeMillis() + ((Double) options.get("interval")).longValue());
    intervalSchedulerModel.setInterval(((Double) options.get("interval")).longValue()); // on iOS we cannot change interval

    SchedulerImpl scheduler = new SchedulerImpl(intervalSchedulerModel);

    SchedulersManagerProxy.getInstance(getReactApplicationContext().getApplicationContext()).addScheduler(
      scheduler,
      (String id) -> {
        if (id == null) {
          promise.reject(new UnableToScheduleException());
          return false;
        }
        promise.resolve(id);
        return true;
      }
    );
  }

  @ReactMethod
  public void scheduleNotificationWithCalendar(final ReadableMap data, final ReadableMap optionsMap, final Promise promise) {
    HashMap<String, Object> options = optionsMap.toHashMap();
    int notificationId = Math.abs(new Random().nextInt());
    HashMap<String, Object> hashMap = data.toHashMap();
    if (data.hasKey("categoryId")) {
      hashMap.put("categoryId", getScopedIdIfNotDetached(data.getString("categoryId")));
    }
    HashMap<String, Object> details = new HashMap<>();
    details.put("data", hashMap);
    String experienceId;

    try {
      experienceId = mManifest.getStableLegacyID();
      details.put("experienceId", experienceId);
    } catch (Exception e) {
      promise.reject(new Exception("Requires Experience Id"));
      return;
    }

    Cron cron = createCronInstance(options);

    CalendarSchedulerModel calendarSchedulerModel = new CalendarSchedulerModel();
    calendarSchedulerModel.setExperienceId(experienceId);
    calendarSchedulerModel.setNotificationId(notificationId);
    calendarSchedulerModel.setDetails(details);
    calendarSchedulerModel.setRepeat(options.containsKey("repeat") && (Boolean) options.get("repeat"));
    calendarSchedulerModel.setCalendarData(cron.asString());

    SchedulerImpl scheduler = new SchedulerImpl(calendarSchedulerModel);

    SchedulersManagerProxy.getInstance(getReactApplicationContext().getApplicationContext()).addScheduler(
      scheduler,
      (String id) -> {
        if (id == null) {
          promise.reject(new UnableToScheduleException());
          return false;
        }
        promise.resolve(id);
        return true;
      }
    );
  }

}

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsModule.kt
// Copyright 2015-present 650 Industries. All rights reserved.
package versioned.host.exp.exponent.modules.api.notifications

import com.facebook.react.bridge.*
import com.google.firebase.iid.FirebaseInstanceId
import host.exp.exponent.Constants
import host.exp.exponent.ExponentManifest
import host.exp.exponent.analytics.EXL.e
import host.exp.exponent.di.NativeModuleDepsProvider
import host.exp.exponent.kernel.ExperienceKey
import host.exp.exponent.network.ExponentNetwork
import host.exp.exponent.notifications.ExponentNotificationManager
import host.exp.exponent.notifications.NotificationActionCenter
import host.exp.exponent.notifications.NotificationConstants
import host.exp.exponent.notifications.NotificationHelper
import host.exp.exponent.notifications.NotificationHelper.TokenListener
import host.exp.exponent.notifications.exceptions.UnableToScheduleException
import host.exp.exponent.notifications.helpers.ExpoCronParser.createCronInstance
import host.exp.exponent.notifications.managers.SchedulersManagerProxy
import host.exp.exponent.notifications.schedulers.CalendarSchedulerModel
import host.exp.exponent.notifications.schedulers.IntervalSchedulerModel
import host.exp.exponent.notifications.schedulers.SchedulerImpl
import host.exp.exponent.storage.ExponentSharedPreferences
import java.io.Serializable
import java.util.*
import javax.inject.Inject
import kotlin.math.abs

class NotificationsModule(
  reactContext: ReactApplicationContext,
  private val experienceKey: ExperienceKey,
  private val experienceStableLegacyId: String,
) : ReactContextBaseJavaModule(reactContext) {
  @Inject
  lateinit var exponentSharedPreferences: ExponentSharedPreferences

  @Inject
  lateinit var exponentManifest: ExponentManifest

  @Inject
  lateinit var exponentNetwork: ExponentNetwork

  override fun getName(): String {
    return "ExponentNotifications"
  }

  @ReactMethod
  fun createCategoryAsync(categoryIdParam: String?, actions: ReadableArray, promise: Promise) {
    val categoryId = getScopedIdIfNotDetached(categoryIdParam)

    val newActions = mutableListOf<MutableMap<String?, Any?>>()
    for (actionObject in actions.toArrayList()) {
      if (actionObject is Map<*, *>) {
        val action = actionObject as Map<String?, Any?>
        newActions.add(action.toMutableMap())
      }
    }

    NotificationActionCenter.putCategory(categoryId, newActions)
    promise.resolve(null)
  }

  @ReactMethod
  fun deleteCategoryAsync(categoryIdParam: String?, promise: Promise) {
    val categoryId = getScopedIdIfNotDetached(categoryIdParam)
    NotificationActionCenter.removeCategory(categoryId)
    promise.resolve(null)
  }

  private fun getScopedIdIfNotDetached(categoryId: String?): String? {
    if (!Constants.isStandaloneApp()) {
      val experienceScopeKey = experienceKey.scopeKey
      return "$experienceScopeKey:$categoryId"
    }
    return categoryId
  }

  @ReactMethod
  fun getDevicePushTokenAsync(config: ReadableMap?, promise: Promise) {
    if (!Constants.isStandaloneApp()) {
      promise.reject("getDevicePushTokenAsync is only accessible within standalone applications")
    }
    try {
      if (Constants.FCM_ENABLED) {
        val token = FirebaseInstanceId.getInstance().token
        if (token == null) {
          promise.reject("FCM token has not been set")
        } else {
          val params = Arguments.createMap().apply {
            putString("type", "fcm")
            putString("data", token)
          }
          promise.resolve(params)
        }
      } else {
        promise.reject(
          "ERR_NOTIFICATIONS_FCM_NOT_ENABLED",
          "FCM must be enabled in order to get the device push token"
        )
      }
    } catch (e: Exception) {
      e(TAG, e.message)
      promise.reject(e.message)
    }
  }

  @ReactMethod
  fun getExponentPushTokenAsync(promise: Promise) {
    val uuid = exponentSharedPreferences.getUUID()
    if (uuid == null) {
      // This should have been set by ExponentNotificationIntentService via
      // ExpoFcmMessagingService#onNewToken -> FcmRegistrationIntentService.registerForeground -> ExponentNotificationIntentService#onHandleIntent
      // (#onNewToken is supposed to be called when a token is generated after app install, see
      // https://developers.google.com/android/reference/com/google/firebase/messaging/FirebaseMessagingService#onNewToken(java.lang.String)).
      // If it hasn't been set, the app probably couldn't register at FCM (invalid configuration?).
      promise.reject(
        "E_GET_PUSH_TOKEN_FAILED",
        "Couldn't get push token on device. Check that your FCM configuration is valid."
      )
      return
    }

    try {
      val experienceId = experienceStableLegacyId
      NotificationHelper.getPushNotificationToken(
        uuid,
        experienceId,
        exponentNetwork,
        exponentSharedPreferences,
        object : TokenListener {
          override fun onSuccess(token: String) {
            promise.resolve(token)
          }

          override fun onFailure(e: Exception) {
            promise.reject(
              "E_GET_PUSH_TOKEN_FAILED",
              "Couldn't get push token for device. Check that your FCM configuration is valid.",
              e
            )
          }
        }
      )
    } catch (e: Exception) {
      promise.reject(
        "E_GET_PUSH_TOKEN_FAILED",
        "Couldn't get push token for device. Check that your FCM configuration is valid.",
        e
      )
    }
  }

  @ReactMethod
  fun createChannel(channelId: String, data: ReadableMap, promise: Promise) {
    val channelName = if (data.hasKey(NotificationConstants.NOTIFICATION_CHANNEL_NAME)) {
      data.getString(NotificationConstants.NOTIFICATION_CHANNEL_NAME)!!
    } else {
      promise.reject("E_FAILED_CREATING_CHANNEL", "Requires channel name")
      return
    }

    try {
      NotificationHelper.createChannel(
        reactApplicationContext,
        experienceKey,
        channelId,
        channelName,
        data.toHashMap()
      )
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject("E_FAILED_CREATING_CHANNEL", "Could not create channel", e)
    }
  }

  @ReactMethod
  fun deleteChannel(channelId: String, promise: Promise) {
    try {
      NotificationHelper.deleteChannel(
        reactApplicationContext,
        experienceKey,
        channelId
      )
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject("E_FAILED_DELETING_CHANNEL", "Could not delete channel", e)
    }
  }

  @ReactMethod
  fun presentLocalNotification(data: ReadableMap, promise: Promise) {
    presentLocalNotificationWithChannel(data, null, promise)
  }

  @ReactMethod
  fun presentLocalNotificationWithChannel(
    data: ReadableMap,
    legacyChannelData: ReadableMap?,
    promise: Promise
  ) {
    val details = hashMapOf<String, Serializable>(
      "data" to data.toHashMap().apply {
        if (data.hasKey("categoryId")) {
          this["categoryId"] = getScopedIdIfNotDetached(data.getString("categoryId"))
        }
      },
      NotificationConstants.NOTIFICATION_EXPERIENCE_ID_KEY to experienceKey.scopeKey,
      NotificationConstants.NOTIFICATION_EXPERIENCE_SCOPE_KEY_KEY to experienceKey.scopeKey
    )

    if (legacyChannelData != null) {
      val channelId = data.getString("channelId")
      if (channelId == null) {
        promise.reject(
          "E_FAILED_PRESENTING_NOTIFICATION",
          "legacyChannelData was nonnull with no channelId"
        )
        return
      }
      NotificationHelper.maybeCreateLegacyStoredChannel(
        reactApplicationContext,
        experienceKey,
        channelId,
        legacyChannelData.toHashMap()
      )
    }

    val notificationId = Random().nextInt()

    NotificationHelper.showNotification(
      reactApplicationContext,
      notificationId,
      details,
      exponentManifest,
      object : NotificationHelper.Listener {
        override fun onSuccess(id: Int) {
          promise.resolve(id)
        }

        override fun onFailure(e: Exception) {
          promise.reject(e)
        }
      }
    )
  }

  @ReactMethod
  fun scheduleLocalNotification(data: ReadableMap, options: ReadableMap, promise: Promise) {
    scheduleLocalNotificationWithChannel(data, options, null, promise)
  }

  @ReactMethod
  fun scheduleLocalNotificationWithChannel(
    data: ReadableMap,
    options: ReadableMap,
    legacyChannelData: ReadableMap?,
    promise: Promise
  ) {
    if (legacyChannelData != null) {
      val channelId = data.getString("channelId")
      if (channelId == null) {
        promise.reject(
          "E_FAILED_PRESENTING_NOTIFICATION",
          "legacyChannelData was nonnull with no channelId or no experienceId"
        )
        return
      }
      NotificationHelper.maybeCreateLegacyStoredChannel(
        reactApplicationContext,
        experienceKey,
        channelId,
        legacyChannelData.toHashMap()
      )
    }

    val notificationId = Random().nextInt()

    val hashMap = data.toHashMap().apply {
      if (data.hasKey("categoryId")) {
        this["categoryId"] = getScopedIdIfNotDetached(data.getString("categoryId"))
      }
    }

    NotificationHelper.scheduleLocalNotification(
      reactApplicationContext,
      notificationId,
      hashMap,
      options.toHashMap(),
      experienceKey,
      object : NotificationHelper.Listener {
        override fun onSuccess(id: Int) {
          promise.resolve(id)
        }

        override fun onFailure(e: Exception) {
          promise.reject(e)
        }
      }
    )
  }

  @ReactMethod
  fun dismissNotification(notificationId: Int, promise: Promise) {
    val manager = ExponentNotificationManager(reactApplicationContext)
    manager.cancel(experienceKey, notificationId)
    promise.resolve(true)
  }

  @ReactMethod
  fun dismissAllNotifications(promise: Promise) {
    val manager = ExponentNotificationManager(reactApplicationContext)
    manager.cancelAll(experienceKey)
    promise.resolve(true)
  }

  @ReactMethod
  fun cancelScheduledNotificationAsync(notificationId: Int, promise: Promise) {
    try {
      val manager = ExponentNotificationManager(reactApplicationContext)
      manager.cancelScheduled(experienceKey, notificationId)
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun cancelScheduledNotificationWithStringIdAsync(id: String?, promise: Promise) {
    try {
      SchedulersManagerProxy.getInstance(reactApplicationContext.applicationContext)
        .removeScheduler(id)
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun cancelAllScheduledNotificationsAsync(promise: Promise) {
    try {
      val manager = ExponentNotificationManager(reactApplicationContext)
      manager.cancelAllScheduled(experienceKey)
      SchedulersManagerProxy.getInstance(reactApplicationContext.applicationContext)
        .removeAll(experienceKey)
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun scheduleNotificationWithTimer(data: ReadableMap, optionsMap: ReadableMap, promise: Promise) {
    val options = optionsMap.toHashMap()
    val notificationId = abs(Random().nextInt())

    val details = hashMapOf<String, Any>(
      "data" to data.toHashMap().apply {
        if (data.hasKey("categoryId")) {
          this["categoryId"] = getScopedIdIfNotDetached(data.getString("categoryId"))
        }
      },
      NotificationConstants.NOTIFICATION_EXPERIENCE_ID_KEY to experienceKey.scopeKey,
      NotificationConstants.NOTIFICATION_EXPERIENCE_SCOPE_KEY_KEY to experienceKey.scopeKey,
    )

    val intervalSchedulerModel = IntervalSchedulerModel().apply {
      experienceScopeKey = experienceKey.scopeKey
      this.notificationId = notificationId
      setDetailsFromMap(details)
      isRepeat = options.containsKey("repeat") && (options["repeat"] as Boolean?)!!
      scheduledTime = System.currentTimeMillis() + (options["interval"] as Double?)!!.toLong()
      interval = (options["interval"] as Double?)!!.toLong() // on iOS we cannot change interval
    }
    val scheduler = SchedulerImpl(intervalSchedulerModel)
    SchedulersManagerProxy.getInstance(reactApplicationContext.applicationContext).addScheduler(
      scheduler
    ) { id: String? ->
      if (id == null) {
        promise.reject(UnableToScheduleException())
        return@addScheduler false
      }
      promise.resolve(id)
      true
    }
  }

  @ReactMethod
  fun scheduleNotificationWithCalendar(
    data: ReadableMap,
    optionsMap: ReadableMap,
    promise: Promise
  ) {
    val options = optionsMap.toHashMap()
    val notificationId = abs(Random().nextInt())

    val details = hashMapOf<String, Any>(
      "data" to data.toHashMap().apply {
        if (data.hasKey("categoryId")) {
          this["categoryId"] = getScopedIdIfNotDetached(data.getString("categoryId"))
        }
      },
      NotificationConstants.NOTIFICATION_EXPERIENCE_ID_KEY to experienceKey.scopeKey,
      NotificationConstants.NOTIFICATION_EXPERIENCE_SCOPE_KEY_KEY to experienceKey.scopeKey
    )
    val cron = createCronInstance(options)
    val calendarSchedulerModel = CalendarSchedulerModel().apply {
      experienceScopeKey = experienceKey.scopeKey
      this.notificationId = notificationId
      setDetailsFromMap(details)
      isRepeat = options.containsKey("repeat") && (options["repeat"] as Boolean?)!!
      calendarData = cron.asString()
    }
    val scheduler = SchedulerImpl(calendarSchedulerModel)
    SchedulersManagerProxy.getInstance(reactApplicationContext.applicationContext).addScheduler(
      scheduler
    ) { id: String? ->
      if (id == null) {
        promise.reject(UnableToScheduleException())
        return@addScheduler false
      }
      promise.resolve(id)
      true
    }
  }

  companion object {
    private val TAG = NotificationsModule::class.java.simpleName
  }

  init {
    NativeModuleDepsProvider.instance.inject(NotificationsModule::class.java, this)
  }
}
